pkg/metrics/collector.go

```diff
@@ -0,0 +1,91 @@
+package metrics
+
+import (
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/kube-parrot/pkg/bgp"
+	"net"
+)
+
+var sessionStati = []string{"idle", "connect", "active", "opensent", "openconfirm", "established"}
+
+type collector struct {
+	nodeName  string
+	neighbors []*net.IP
+	bgpServer *bgp.Server
+
+	bgpNeighborsSessionStatusMetric,
+	bgpNeighborAdvertisedRouteCountTotalMetric *prometheus.Desc
+}
+
+func RegisterCollector(nodeName string, neighbors []*net.IP, bgpServer *bgp.Server) {
+	prometheus.MustRegister(
+		newCollector(nodeName, neighbors, bgpServer),
+	)
+}
+
+func newCollector(nodeName string, neighbors []*net.IP, bgpServer *bgp.Server) *collector {
+	return &collector{
+		nodeName:  nodeName,
+		neighbors: neighbors,
+		bgpServer: bgpServer,
+		bgpNeighborsSessionStatusMetric: prometheus.NewDesc(
+			"kube_parrot_bgp_neighbor_session_status",
+			"Session status of BGP neighbors.",
+			[]string{"node", "neighbor", "status"},
+			nil,
+		),
+		bgpNeighborAdvertisedRouteCountTotalMetric: prometheus.NewDesc(
+			"kube_parrot_bgp_neighbor_advertised_route_count_total",
+			"Total count of advertised routes to BGP neighbor.",
+			[]string{"node", "neighbor"},
+			nil,
+		),
+	}
+}
+
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.bgpNeighborsSessionStatusMetric
+	ch <- c.bgpNeighborAdvertisedRouteCountTotalMetric
+}
+
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	for _, neighbor := range c.neighbors {
+		neighborList, err := c.bgpServer.GetNeighbor(neighbor.String())
+		if err != nil {
+			glog.Infof("failed to get session status for BGP neighbor: %v", err)
+			continue
```

Comment:
Maybe you tested this but wouldn’t in case of an error the metric just stay at the previous value?

Not that the server returns an error when the neighbor doesn’t exist and we never notice the problem. Or the bgp server is somehow crashed and doesn’t answer at all anymore.

If we are unclear when exactly this errors, I would propose to set the gauge to -1.  Then we can even distinguish between parrot broken (-1) and aci broken (0) and avi degraded (1).

Reply:
Introduced a new metric for errors. Could be extended in separate PR.